File: app/llm/openai_adapter.py
"""OpenAI 适配器实现（非流式 + SSE 流式）。"""

from __future__ import annotations

import json
from typing import Any, Dict, Iterator, Optional

from app.llm.base import (
    BaseAdapter,
    StreamTransport,
    Transport
)
from app.llm.types import (
    DocumentPart,
    ImageBlock,
    ImagePart,
    LLMMessage,
    LLMRequest,
    LLMResponse,
    LLMTool,
    LLMUsage,
    MessageContent,
    ParsedResponse,
    StreamChunk,
    TextBlock,
    TextPart,
    ToolCallBlock,
    content_to_text,
)


class OpenAIAdapter(BaseAdapter):
    """OpenAI LLM 适配器。"""

    def __init__(self, api_key: str, base_url: str, transport: Transport, timeout_sec: int = 60) -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip('/')
        self._transport = transport
        self._timeout_sec = timeout_sec

    # ── 非流式 ────────────────────────────────────────────────────────────

    def complete(self, req: LLMRequest) -> LLMResponse:
        """统一补全接口（OpenAI Chat Completions 风格）。"""
        payload = self._build_payload(req, stream=False)
        url = f"{self._base_url}/v1/chat/completions"
        resp = self._transport.post(url, headers=self._headers(), json=payload, timeout=self._timeout_sec)
        return LLMResponse(
            text=_extract_openai_text(resp),
            raw=resp,
            usage=_extract_openai_usage(resp),
        )

    def parse_response(self, response: LLMResponse) -> ParsedResponse:
        """解析 OpenAI 响应为统一内容块。"""
        raw = response.raw or {}
        message = _extract_openai_message(raw)
        blocks: list = []
        tool_calls: list[ToolCallBlock] = []
        images: list[ImageBlock] = []

        for text in _extract_openai_text_parts(message):
            blocks.append(TextBlock(type='text', text=text))

        for img in _extract_openai_image_parts(message):
            blocks.append(img)
            images.append(img)

        for call in _extract_openai_tool_calls(message):
            blocks.append(call)
            tool_calls.append(call)

        text = '\n'.join(b.text for b in blocks if isinstance(b, TextBlock)).strip()
        return ParsedResponse(text=text, blocks=blocks, tool_calls=tool_calls, images=images, raw=raw, usage=response.usage)

    # ── 流式 ──────────────────────────────────────────────────────────────

    def stream(self, req: LLMRequest) -> Iterator[StreamChunk]:
        """SSE 流式补全，逐 token 产出 StreamChunk。

        要求 transport 实现 StreamTransport 协议（如 HttpxTransport）。
        """
        if not isinstance(self._transport, StreamTransport):
            yield from super().stream(req)
            return

        payload = self._build_payload(req, stream=True)
        url = f"{self._base_url}/v1/chat/completions"

        # 聚合 tool_call 增量（按 index）
        tool_call_buffers: dict[int, dict[str, Any]] = {}

        for raw_line in self._transport.stream_post(url, self._headers(), payload, self._timeout_sec):
            try:
                data = json.loads(raw_line)
            except json.JSONDecodeError:
                continue

            choices = data.get('choices') or []
            if not choices:
                # 最后一个 chunk 可能只含 usage
                usage_data = data.get('usage')
                if usage_data:
                    yield StreamChunk(is_done=True, usage=_parse_usage(usage_data))
                continue

            choice = choices[0]
            delta = choice.get('delta') or {}
            finish_reason = choice.get('finish_reason')

            # 正文 / 推理 / 图片增量分开处理，避免 reasoning_content 混入最终回复文本。
            for text_delta in _iter_openai_content_delta(delta):
                yield StreamChunk(text_delta=text_delta)

            for reasoning_delta in _iter_openai_reasoning_delta(delta):
                yield StreamChunk(reasoning_delta=reasoning_delta)

            for image in _iter_openai_delta_images(delta):
                yield StreamChunk(image=image)

            # tool_call 增量
            for tc_delta in delta.get('tool_calls') or []:
                idx = tc_delta.get('index', 0)
                buf = tool_call_buffers.setdefault(idx, {'id': '', 'name': '', 'arguments': ''})
                if tc_delta.get('id'):
                    buf['id'] = tc_delta['id']
                fn = tc_delta.get('function') or {}
                if fn.get('name'):
                    buf['name'] += fn['name']
                if fn.get('arguments'):
                    buf['arguments'] += fn['arguments']
                yield StreamChunk(tool_call_delta={'index': idx, **buf})

            if finish_reason:
                usage_data = data.get('usage')
                yield StreamChunk(
                    is_done=True,
                    finish_reason=finish_reason,
                    usage=_parse_usage(usage_data) if usage_data else None,
                )
                # don't return: let the loop continue so the trailing usage-only
                # chunk (choices=[]) can be captured when stream_options is used

    # ── 内部工具 ──────────────────────────────────────────────────────────

    def _headers(self) -> Dict[str, str]:
        return {
            'Authorization': f"Bearer {self._api_key}",
            'Content-Type': 'application/json',
        }

    def _build_payload(self, req: LLMRequest, stream: bool) -> Dict[str, Any]:
        messages = _merge_system_prompt(req)
        payload: Dict[str, Any] = {
            'model': req.model,
            'messages': _serialize_messages_openai(messages),
            'stream': stream,
        }
        if stream:
            payload['stream_options'] = {'include_usage': True}
        if req.tools:
            payload['tools'] = _map_openai_tools(req.tools)
        if req.temperature is not None:
            payload['temperature'] = req.temperature
        if req.max_tokens is not None:
            payload['max_tokens'] = req.max_tokens
        if req.top_p is not None:
            payload['top_p'] = req.top_p
        if req.stop is not None:
            payload['stop'] = req.stop
        return payload


# ── 私有解析函数 ──────────────────────────────────────────────────────────

def _parse_usage(usage: Dict[str, Any]) -> LLMUsage:
    return LLMUsage(
        prompt_tokens=usage.get('prompt_tokens'),
        completion_tokens=usage.get('completion_tokens'),
        total_tokens=usage.get('total_tokens'),
    )


def _iter_openai_content_delta(delta: Dict[str, Any]) -> Iterator[str]:
    """从流式 delta.content 中提取正文增量。

    兼容形态：
    - {"content": "..."}
    - {"content": [{"type":"text","text":"..."}]}
    """
    value = delta.get('content')
    if isinstance(value, str):
        if value:
            yield value
        return

    if isinstance(value, dict):
        text = value.get('text') or value.get('content') or ''
        if text:
            yield text
        return

    if not isinstance(value, list):
        return

    for item in value:
        if not isinstance(item, dict):
            continue
        item_type = item.get('type')
        if item_type in ('text', 'output_text'):
            text = item.get('text') or item.get('content') or ''
            if text:
                yield text


def _iter_openai_reasoning_delta(delta: Dict[str, Any]) -> Iterator[str]:
    """从流式 delta.reasoning_content 中提取推理增量。"""
    value = delta.get('reasoning_content')
    if isinstance(value, str):
        if value:
            yield value
        return

    if isinstance(value, dict):
        text = value.get('text') or value.get('content') or ''
        if text:
            yield text
        return

    if not isinstance(value, list):
        return

    for item in value:
        if not isinstance(item, dict):
            continue
        item_type = item.get('type')
        if item_type in ('text', 'reasoning_text', 'output_text'):
            text = item.get('text') or item.get('content') or ''
            if text:
                yield text


def _iter_openai_delta_images(delta: Dict[str, Any]) -> Iterator[ImageBlock]:
    """从流式 delta 中提取图片块。"""
    content_delta = delta.get('content')
    if not isinstance(content_delta, list):
        return

    for item in content_delta:
        if not isinstance(item, dict) or item.get('type') != 'image_url':
            continue
        url = (item.get('image_url') or {}).get('url', '')
        if url.startswith('data:'):
            media_type, data = _parse_data_url(url)
            yield ImageBlock(type='image', media_type=media_type, source_type='base64', data=data)
        elif url:
            yield ImageBlock(type='image', media_type='', source_type='url', data=url)


def _extract_openai_text(resp: Dict[str, Any]) -> str:
    choices = resp.get('choices') or []
    if not choices:
        return ''
    message = choices[0].get('message') or {}
    return message.get('content') or ''


def _extract_openai_usage(resp: Dict[str, Any]) -> Optional[LLMUsage]:
    usage = resp.get('usage') or {}
    return _parse_usage(usage) if usage else None


def _extract_openai_message(resp: Dict[str, Any]) -> Dict[str, Any]:
    choices = resp.get('choices') or []
    if not choices:
        return {}
    return choices[0].get('message') or {}


def _extract_openai_text_parts(message: Dict[str, Any]) -> list[str]:
    content = message.get('content')
    if isinstance(content, str):
        return [content] if content else []
    if isinstance(content, list):
        return [item.get('text') or '' for item in content if item.get('type') == 'text' and item.get('text')]
    return []


def _parse_data_url(url: str) -> tuple[str, str]:
    """解析 data URL，返回 (media_type, base64_data)。"""
    # data:<media_type>;base64,<data>
    try:
        header, data = url.split(',', 1)
        media_type = header.split(':')[1].split(';')[0]
        return media_type, data
    except Exception:
        return '', url


def _extract_openai_image_parts(message: Dict[str, Any]) -> list[ImageBlock]:
    content = message.get('content')
    if not isinstance(content, list):
        return []
    images: list[ImageBlock] = []
    for item in content:
        if item.get('type') != 'image_url':
            continue
        url = (item.get('image_url') or {}).get('url', '')
        if url.startswith('data:'):
            media_type, data = _parse_data_url(url)
            images.append(ImageBlock(type='image', media_type=media_type, source_type='base64', data=data))
        else:
            images.append(ImageBlock(type='image', media_type='', source_type='url', data=url))
    return images


def _extract_openai_tool_calls(message: Dict[str, Any]) -> list[ToolCallBlock]:
    blocks: list[ToolCallBlock] = []
    for call in message.get('tool_calls') or []:
        fn = call.get('function') or {}
        args_raw = fn.get('arguments') or ''
        try:
            args = json.loads(args_raw) if args_raw else {}
        except json.JSONDecodeError:
            args = {'_raw_arguments': args_raw}
        blocks.append(ToolCallBlock(
            type='tool_call',
            id=call.get('id') or '',
            name=fn.get('name') or '',
            input=args,
            tool_type=call.get('type') or 'function',
            raw=call,
        ))
    return blocks


def _content_to_text(content: MessageContent) -> str:
    return content_to_text(content)


def _content_to_openai_blocks(content: MessageContent) -> 'list[dict] | str':
    """将 MessageContent 转为 OpenAI content blocks。
    纯字符串直接返回（向后兼容），part 列表逐项转换。
    DocumentPart 转为 data URL（OpenAI 不支持文档原生格式）。
    """
    if isinstance(content, str):
        return content
    blocks: list[dict] = []
    for part in content:
        if isinstance(part, TextPart):
            if part.text:
                blocks.append({"type": "text", "text": part.text})
        elif isinstance(part, ImagePart):
            url = part.data if part.source_type == "url" else f"data:{part.media_type};base64,{part.data}"
            blocks.append({"type": "image_url", "image_url": {"url": url}})
        elif isinstance(part, DocumentPart):
            url = f"data:{part.media_type};base64,{part.data}"
            blocks.append({"type": "image_url", "image_url": {"url": url}})
    return blocks or ""


def _serialize_messages_openai(messages: list[LLMMessage]) -> list[dict]:
    """将内部 LLMMessage 列表序列化为 OpenAI Chat Completions API 格式。

    - role="assistant" + tool_calls → tool_calls 数组（arguments JSON 序列化）
    - role="tool" + tool_call_id → {role:"tool", tool_call_id:..., content:...}
    - role="tool" 且 tool_call_id 为空 → 降级为 role="user" 文本（历史记忆回放）
    - 其他 role → content 经 _content_to_openai_blocks 转换（支持多模态）
    """
    result: list[dict] = []
    for m in messages:
        if m.role == "assistant" and m.tool_calls:
            entry: dict = {"role": "assistant"}
            if m.reasoning_content:
                entry["reasoning_content"] = m.reasoning_content
            text = _content_to_text(m.content)
            if text:
                entry["content"] = text
            entry["tool_calls"] = [
                {
                    "id": tc["id"],
                    "type": "function",
                    "function": {
                        "name": tc["name"],
                        "arguments": json.dumps(tc["input"], ensure_ascii=False),
                    },
                }
                for tc in m.tool_calls
            ]
            result.append(entry)

        elif m.role == "tool":
            if m.tool_call_id:
                result.append({
                    "role": "tool",
                    "tool_call_id": m.tool_call_id,
                    "content": _content_to_text(m.content),
                })
            else:
                result.append({"role": "user", "content": _content_to_text(m.content)})

        else:
            result.append({"role": m.role, "content": _content_to_openai_blocks(m.content)})

    return result


def _merge_system_prompt(req: LLMRequest) -> list[LLMMessage]:
    messages = list(req.messages)
    if not req.system_prompt:
        return messages
    for i, m in enumerate(messages):
        if m.role == 'system':
            existing = _content_to_text(m.content)
            messages[i] = LLMMessage(role='system', content=f"{req.system_prompt}\n{existing}".strip())
            return messages
    return [LLMMessage(role='system', content=req.system_prompt)] + messages


def _map_openai_tools(tools: list[LLMTool]) -> list[dict]:
    mapped: list[dict] = []
    for tool in tools:
        if tool.type != 'function':
            raise ValueError(f'OpenAI 仅支持 function 工具，当前: {tool.type}')
        fn: dict = {'name': tool.name, 'parameters': tool.input_schema.to_dict()}
        if tool.description:
            fn['description'] = tool.description
        mapped.append({'type': 'function', 'function': fn})
    return mapped

File: miniagents-assembled/src/main/java/com/codex/miniagents/llm/adapter/OpenAiAdapter.java
package com.codex.miniagents.llm.adapter;

import com.codex.miniagents.llm.model.LlmContentBlock;
import com.codex.miniagents.llm.model.LlmMessage;
import com.codex.miniagents.llm.model.LlmRequest;
import com.codex.miniagents.llm.model.LlmResponse;
import com.codex.miniagents.llm.model.LlmTool;
import com.codex.miniagents.llm.model.LlmUsage;
import com.codex.miniagents.llm.model.ParsedResponse;
import com.codex.miniagents.llm.model.StreamChunk;
import com.codex.miniagents.llm.model.TextBlock;
import com.codex.miniagents.llm.model.ToolCallBlock;
import com.codex.miniagents.llm.transport.StreamTransport;
import com.codex.miniagents.llm.transport.Transport;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Queue;

public class OpenAiAdapter extends LlmAdapter {
    private final String apiKey;

    private final String baseUrl;

    private final Transport transport;

    private final int timeoutSec;

    private final ObjectMapper objectMapper = new ObjectMapper();

    public OpenAiAdapter(String apiKey, String baseUrl, Transport transport, int timeoutSec) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.transport = transport;
        this.timeoutSec = timeoutSec;
    }

    @Override
    public LlmResponse complete(LlmRequest req) {
        Map<String, Object> payload = buildPayload(req, false);
        String url = baseUrl + "/v1/chat/completions";
        Map<String, Object> resp = transport.post(url, headers(), payload, timeoutSec);

        return LlmResponse.builder().text(extractOpenAiText(resp)).raw(resp).usage(extractOpenAiUsage(resp)).build();
    }

    @Override
    public ParsedResponse parseResponse(LlmResponse response) {
        Map<String, Object> raw = response.getRaw() == null ? Map.of() : response.getRaw();
        Map<String, Object> message = extractOpenAiMessage(raw);

        List<LlmContentBlock> blocks = new ArrayList<>();
        List<ToolCallBlock> toolCalls = new ArrayList<>();

        for (String text : extractOpenAiTextParts(message)) {
            blocks.add(new TextBlock("text", text));
        }

        for (ToolCallBlock call : extractOpenAiToolCalls(message)) {
            blocks.add(call);
            toolCalls.add(call);
        }

        String text = blocks.stream()
            .filter(TextBlock.class::isInstance)
            .map(TextBlock.class::cast)
            .map(TextBlock::getText)
            .filter(Objects::nonNull)
            .reduce((a, b) -> a + "\n" + b)
            .orElse("")
            .trim();

        return ParsedResponse.builder()
            .text(text)
            .blocks(blocks)
            .toolCalls(toolCalls)
            .raw(raw)
            .usage(response.getUsage())
            .build();
    }

    @Override
    public Iterator<StreamChunk> stream(LlmRequest req) {
        if (!(transport instanceof StreamTransport streamTransport)) {
            return super.stream(req);
        }

        Map<String, Object> payload = buildPayload(req, true);
        String url = baseUrl + "/v1/chat/completions";

        Map<Integer, Map<String, Object>> toolCallBuffers = new HashMap<>();
        Iterator<String> rawIterator = streamTransport.streamPost(url, headers(), payload, timeoutSec);

        return new Iterator<>() {
            private final Queue<StreamChunk> queue = new ArrayDeque<>();

            private boolean done = false;

            @Override
            public boolean hasNext() {
                fillQueue();
                return !queue.isEmpty();
            }

            @Override
            public StreamChunk next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                return queue.poll();
            }

            private void fillQueue() {
                while (queue.isEmpty() && !done && rawIterator.hasNext()) {
                    String rawLine = rawIterator.next();
                    Map<String, Object> data;
                    try {
                        data = objectMapper.readValue(rawLine, new TypeReference<>() {});
                    } catch (Exception e) {
                        continue;
                    }

                    List<Map<String, Object>> choices = getList(data.get("choices"));
                    if (choices.isEmpty()) {
                        Object usageData = data.get("usage");
                        if (usageData instanceof Map<?, ?> usageMap) {
                            queue.add(StreamChunk.builder().done(true).usage(parseUsage(castMap(usageMap))).build());
                            done = true;
                        }
                        continue;
                    }

                    Map<String, Object> choice = choices.get(0);
                    Map<String, Object> delta = getMap(choice.get("delta"));
                    Object finishReason = choice.get("finish_reason");

                    String textDelta = asString(delta.get("content"));
                    if (!textDelta.isEmpty()) {
                        queue.add(StreamChunk.builder().textDelta(textDelta).build());
                    }

                    for (Map<String, Object> tcDelta : getList(delta.get("tool_calls"))) {
                        int idx = getInt(tcDelta.get("index"), 0);
                        Map<String, Object> buf = toolCallBuffers.computeIfAbsent(idx, k -> {
                            Map<String, Object> m = new LinkedHashMap<>();
                            m.put("id", "");
                            m.put("name", "");
                            m.put("arguments", "");
                            return m;
                        });

                        String id = asString(tcDelta.get("id"));
                        if (!id.isEmpty()) {
                            buf.put("id", id);
                        }

                        Map<String, Object> fn = getMap(tcDelta.get("function"));
                        String name = asString(fn.get("name"));
                        if (!name.isEmpty()) {
                            buf.put("name", asString(buf.get("name")) + name);
                        }

                        String arguments = asString(fn.get("arguments"));
                        if (!arguments.isEmpty()) {
                            buf.put("arguments", asString(buf.get("arguments")) + arguments);
                        }

                        Map<String, Object> deltaObj = new LinkedHashMap<>();
                        deltaObj.put("index", idx);
                        deltaObj.putAll(buf);
                        queue.add(StreamChunk.builder().toolCallDelta(deltaObj).build());
                    }

                    if (finishReason != null) {
                        Map<String, Object> usageData = getMap(data.get("usage"));
                        queue.add(StreamChunk.builder()
                            .done(true)
                            .usage(usageData.isEmpty() ? null : parseUsage(usageData))
                            .build());
                        done = true;
                    }
                }
            }
        };
    }

    private Map<String, String> headers() {
        return Map.of("Authorization", "Bearer " + apiKey, "Content-Type", "application/json");
    }

    private Map<String, Object> buildPayload(LlmRequest req, boolean stream) {
        List<LlmMessage> messages = mergeSystemPrompt(req);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("model", req.getModel());
        payload.put("messages", serializeMessagesOpenAi(messages));
        payload.put("stream", stream);
        if (req.getTools() != null && !req.getTools().isEmpty()) {
            payload.put("tools", mapOpenAiTools(req.getTools()));
        }
        if (req.getTemperature() != null) {
            payload.put("temperature", req.getTemperature());
        }
        if (req.getMaxTokens() != null) {
            payload.put("max_tokens", req.getMaxTokens());
        }
        if (req.getTopP() != null) {
            payload.put("top_p", req.getTopP());
        }
        if (req.getStop() != null) {
            payload.put("stop", req.getStop());
        }
        return payload;
    }

    private List<Map<String, Object>> serializeMessagesOpenAi(List<LlmMessage> messages) {
        List<Map<String, Object>> result = new ArrayList<>();
        for (LlmMessage message : messages) {
            if (message == null) {
                continue;
            }
            if ("assistant".equals(message.getRole()) && message.getToolCalls() != null && !message.getToolCalls().isEmpty()) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("role", "assistant");
                if (message.getContent() != null && !message.getContent().isBlank()) {
                    item.put("content", message.getContent());
                }
                List<Map<String, Object>> toolCalls = new ArrayList<>();
                for (ToolCallBlock tc : message.getToolCalls()) {
                    Map<String, Object> function = new LinkedHashMap<>();
                    function.put("name", tc.getName());
                    try {
                        function.put("arguments", objectMapper.writeValueAsString(tc.getInput() == null ? Map.of() : tc.getInput()));
                    } catch (Exception e) {
                        function.put("arguments", "{}");
                    }
                    Map<String, Object> call = new LinkedHashMap<>();
                    call.put("id", tc.getId());
                    call.put("type", "function");
                    call.put("function", function);
                    toolCalls.add(call);
                }
                item.put("tool_calls", toolCalls);
                result.add(item);
                continue;
            }
            if ("tool".equals(message.getRole())) {
                if (message.getToolCallId() != null && !message.getToolCallId().isBlank()) {
                    Map<String, Object> item = new LinkedHashMap<>();
                    item.put("role", "tool");
                    item.put("tool_call_id", message.getToolCallId());
                    item.put("content", message.getContent() == null ? "" : message.getContent());
                    result.add(item);
                } else {
                    result.add(Map.of(
                        "role", "user",
                        "content", message.getContent() == null ? "" : message.getContent()
                    ));
                }
                continue;
            }
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("role", message.getRole());
            item.put("content", message.getContent());
            result.add(item);
        }
        return result;
    }

    private LlmUsage parseUsage(Map<String, Object> usage) {
        return LlmUsage.builder()
            .promptTokens(getIntegerOrNull(usage.get("prompt_tokens")))
            .completionTokens(getIntegerOrNull(usage.get("completion_tokens")))
            .totalTokens(getIntegerOrNull(usage.get("total_tokens")))
            .build();
    }

    private String extractOpenAiText(Map<String, Object> resp) {
        List<Map<String, Object>> choices = getList(resp.get("choices"));
        if (choices.isEmpty()) {
            return "";
        }
        Map<String, Object> message = getMap(choices.get(0).get("message"));
        return asString(message.get("content"));
    }

    private LlmUsage extractOpenAiUsage(Map<String, Object> resp) {
        Map<String, Object> usage = getMap(resp.get("usage"));
        return usage.isEmpty() ? null : parseUsage(usage);
    }

    private Map<String, Object> extractOpenAiMessage(Map<String, Object> resp) {
        List<Map<String, Object>> choices = getList(resp.get("choices"));
        if (choices.isEmpty()) {
            return Map.of();
        }
        return getMap(choices.get(0).get("message"));
    }

    private List<String> extractOpenAiTextParts(Map<String, Object> message) {
        Object content = message.get("content");
        if (content instanceof String s) {
            return s.isEmpty() ? List.of() : List.of(s);
        }
        if (content instanceof List<?> list) {
            List<String> result = new ArrayList<>();
            for (Object itemObj : list) {
                Map<String, Object> item = castMap(itemObj);
                if ("text".equals(item.get("type")) && item.get("text") != null) {
                    result.add(asString(item.get("text")));
                }
            }
            return result;
        }
        return List.of();
    }

    private List<ToolCallBlock> extractOpenAiToolCalls(Map<String, Object> message) {
        List<ToolCallBlock> blocks = new ArrayList<>();
        for (Map<String, Object> call : getList(message.get("tool_calls"))) {
            Map<String, Object> fn = getMap(call.get("function"));
            String argsRaw = asString(fn.get("arguments"));
            Map<String, Object> args;
            try {
                args = argsRaw.isEmpty()
                    ? new LinkedHashMap<>()
                    : objectMapper.readValue(argsRaw, new TypeReference<>() {});
            } catch (Exception e) {
                args = new LinkedHashMap<>();
                args.put("_raw_arguments", argsRaw);
            }

            blocks.add(new ToolCallBlock("tool_call", asString(call.get("id")), asString(fn.get("name")), args,
                asString(call.getOrDefault("type", "function")), call));
        }
        return blocks;
    }

    private List<LlmMessage> mergeSystemPrompt(LlmRequest req) {
        List<LlmMessage> messages = new ArrayList<>(req.getMessages());
        if (req.getSystemPrompt() == null || req.getSystemPrompt().isBlank()) {
            return messages;
        }

        for (int i = 0; i < messages.size(); i++) {
            LlmMessage m = messages.get(i);
            if ("system".equals(m.getRole())) {
                messages.set(i, LlmMessage.builder()
                    .role("system")
                    .content((req.getSystemPrompt() + "\n" + m.getContent()).trim())
                    .toolCallId(m.getToolCallId())
                    .toolCalls(m.getToolCalls())
                    .build());
                return messages;
            }
        }

        List<LlmMessage> result = new ArrayList<>();
        result.add(LlmMessage.builder().role("system").content(req.getSystemPrompt()).build());
        result.addAll(messages);
        return result;
    }

    private List<Map<String, Object>> mapOpenAiTools(List<LlmTool> tools) {
        List<Map<String, Object>> mapped = new ArrayList<>();
        for (LlmTool tool : tools) {
            if (!"function".equals(tool.getType())) {
                throw new IllegalArgumentException("OpenAI 仅支持 function 工具，当前: " + tool.getType());
            }
            Map<String, Object> fn = new LinkedHashMap<>();
            fn.put("name", tool.getName());
            fn.put("parameters", tool.getInputSchema().toDict());
            if (tool.getDescription() != null && !tool.getDescription().isBlank()) {
                fn.put("description", tool.getDescription());
            }

            Map<String, Object> item = new LinkedHashMap<>();
            item.put("type", "function");
            item.put("function", fn);
            mapped.add(item);
        }
        return mapped;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> castMap(Object obj) {
        return obj instanceof Map<?, ?> map ? (Map<String, Object>) map : new LinkedHashMap<>();
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> getList(Object obj) {
        return obj instanceof List<?> list ? (List<Map<String, Object>>) list : List.of();
    }

    private Map<String, Object> getMap(Object obj) {
        return castMap(obj);
    }

    private String asString(Object obj) {
        return obj == null ? "" : String.valueOf(obj);
    }

    private int getInt(Object obj, int defaultVal) {
        if (obj instanceof Number n) {
            return n.intValue();
        }
        try {
            return obj == null ? defaultVal : Integer.parseInt(String.valueOf(obj));
        } catch (Exception e) {
            return defaultVal;
        }
    }

    private Integer getIntegerOrNull(Object obj) {
        if (obj == null) {
            return null;
        }
        if (obj instanceof Number n) {
            return n.intValue();
        }
        try {
            return Integer.parseInt(String.valueOf(obj));
        } catch (Exception e) {
            return null;
        }
    }
}
